distreqx/bijectors/triangular_linear.py

```diff
@@ -0,0 +1,105 @@
+"""Triangular linear bijector."""
+
+from typing import Tuple
+
+import jax
+import jax.numpy as jnp
+from jaxtyping import Array
+
+from ._bijector import AbstractBijector
+from ._linear import AbstractLinearBijector
+
+
+def _triangular_logdet(matrix: Array) -> Array:
+    """Computes the log absolute determinant of a triangular matrix."""
+    return jnp.sum(jnp.log(jnp.abs(jnp.diag(matrix))))
+
+
+class TriangularLinear(AbstractLinearBijector):
+    """A linear bijector whose weight matrix is triangular.
+
+    The bijector is defined as `f(x) = Ax` where `A` is a DxD triangular matrix.
+
+    The Jacobian determinant can be computed in O(D) as follows:
+
+    log|det J(x)| = log|det A| = sum(log|diag(A)|)
+
+    The inverse is computed in O(D^2) by solving the triangular system `Ax = y`.
+
+    The bijector is invertible if and only if all diagonal elements of `A` are
+    non-zero. It is the responsibility of the user to make sure that this is the
+    case; the class will make no attempt to verify that the bijector is
+    invertible.
+    """
+
+    _matrix: Array
+    _is_lower: bool
+
+    def __init__(self, matrix: Array, is_lower: bool = True):
+        """Initializes a `TriangularLinear` bijector.
+
+        Args:
+        matrix: a square matrix whose triangular part defines `A`. Can also be a
+            batch of matrices. Whether `A` is the lower or upper triangular part of
+            `matrix` is determined by `is_lower`.
+        is_lower: if True, `A` is set to the lower triangular part of `matrix`. If
+            False, `A` is set to the upper triangular part of `matrix`.
+        """
+        if matrix.ndim < 2:
+            raise ValueError(
+                f"`matrix` must have at least 2 dimensions, got {matrix.ndim}."
+            )
+        if matrix.shape[-2] != matrix.shape[-1]:
+            raise ValueError(
+                f"`matrix` must be square; instead, it has shape {matrix.shape[-2:]}."
+            )
+        super().__init__(event_dims=matrix.shape[-1])
+
+        self._matrix = jnp.tril(matrix) if is_lower else jnp.triu(matrix)
+        self._is_lower = is_lower
+
+    @property
+    def matrix(self) -> Array:
+        """The triangular matrix `A` of the transformation."""
+        return self._matrix
+
+    @property
+    def is_lower(self) -> bool:
+        """True if `A` is lower triangular, False if upper triangular."""
+        return self._is_lower
+
+    def forward(self, x: Array) -> Array:
+        """Computes y = f(x)."""
+        return self._matrix @ x
+
+    def forward_log_det_jacobian(self, x: Array) -> Array:
+        """Computes log|det J(f)(x)|."""
+        triangular_logdet = jnp.vectorize(_triangular_logdet, signature="(m,m)->()")
```

Comment:
Generally I prefer `vmap` to `vectorize`, but since there's already vectorizes in the code base I might just go through and change them later